### PR TITLE
Divisive Cluster

### PR DESCRIPTION
**Pull request by tliu68**
#### Reference Issues/PRs

part of #299 

#### What does this implement/fix? Explain your changes.
DivisiveCluster class that recursively clusters data based on a chosen clustering algorithm in a "top-down" fashion

### CODE REVIEW

File: HSBM.py
Comment (bdpedigo) [inline HSBM.py:48]: i'd just call `selection_criteria` and set the default to None, choose the right metric based on the clustering type

File: HSBM.py
Comment (bdpedigo) [inline HSBM.py:50]: this was kinda a hack on my part. guessing we wont want it in the final code

File: HSBM.py
Comment (bdpedigo) [inline HSBM.py:54]: not sure whether to call clusters or components because kmeans it would be clusters. we can discuss

File: HSBM.py
Comment (bdpedigo) [inline HSBM.py:58]: i don't think we want the base, clustering only case to even have to think about this as a possibility

File: HSBM.py
Comment (bdpedigo) [inline HSBM.py:219]: i think we need to get rid of this function, was very hacky on my part, and replace with something recursive.

Comment (bdpedigo) [reply HSBM.py:219]: note that this is kinda specific to the input data

File: HSBM.py
Comment (bdpedigo) [inline HSBM.py:85]: note that this does not use X, which is not what we want. X could be data that was not seen during fit

File: HSBM.py
Comment (alyakin314) [inline HSBM.py:385]: what is planned to be used as a dissimilarity here in place of Minh's T statistic in the original hsbm paper?

File: HSBM.py
Comment (bdpedigo) [inline HSBM.py:97]: predict should not refit the model to the new data that was passed, should only run the prediction step. so will probably want to separate the predict functionality into this function or somewhere else, and then call it from `fit_predict`

File: HSBM.py
Comment (bdpedigo) [inline HSBM.py:109]: is `adj` doing anything here?

File: HSBM.py
Comment (bdpedigo) [inline HSBM.py:131]: it seems odd to me that this is `root.leaves` and not just `self.leaves` but maybe I don't understand the recursion... can you explain?

File: HSBM.py
Comment (bdpedigo) [inline HSBM.py:147]: should this be `min_components`?

File: HSBM.py
Comment (tliu68) [reply HSBM.py:97]: Oh I see what you're saying. Will fix this.

File: HSBM.py
Comment (tliu68) [reply HSBM.py:131]: So the way I imagine it should work is this: go through all leaves in the current tree and pick the first node that satisfy some conditions which are 1) not too few samples to be fitted w/ AGMM or not fewer samples than min_split; 2) its depth not greater than max_level; 3) not yet fitted. I think `root.leaves` makes sure that it go through all the leaves in the overall tree. But I can also see how to iterate w/ `self.leaves`. Once a node gets `k=1`, the algorithm checks its siblings or moves up and check some ancestor's siblings. What do you prefer?

File: HSBM.py
Comment (tliu68) [reply HSBM.py:147]: Sorry I don't follow. Which line is it?

Comment (bdpedigo) [reply HSBM.py:147]: line 147 you always set min_components to 1, is that desired?

Comment (tliu68) [reply HSBM.py:147]: I'm setting it as 1 because I don't really see why someone would want a higher threshold for it: 
i) if that's not together w/ settings on max_level or min_split, the algorithm will keep splitting (until too few samples left); 
ii) even if it is indeed, I think the person would have a very clear vision for the data and really want the original data (& all following sub-clusters) to be recursively clustered until a specific level or size.

But I'm kind of hesitant about doing this for level 0 as well. Maybe I should change it so that it allows customized min_comp at level 0. 

And maybe there are other cases where someone does have the need to customize I didn't think of. What do you imagine?

File: HSBM.py
Comment (bdpedigo) [reply HSBM.py:109]: i feel like this is not the formatting/inheritance we discussed the last time we talked about it, do you agree?

Comment (bdpedigo) [reply HSBM.py:109]: in the sense that here, _RecursiveGraphCluster inherits from this, which is not what I thought we planned?

**Comment by bdpedigo** [comment]
@tliu68 i'm not sure if you're waiting for me (to review more) or im waiting for you on this PR. lmk which it is? 

also, wondering if we can break this up into smaller PRs to make reviewing more manageable. i think i wrote something about how to break this up in that google doc we discussed on. what are your thoughts?

**Comment by bdpedigo** [comment]
@tliu68 I am going to answer some of the questions you sent me here so everything is in one place and public.

>  (minor: I named the RecursiveCluster file as "rclust". It seems to follow the convention?)

Yes that's fine, we should change the convention someday, but it is the current convention. 

> test: apart from the very basic part (input test, test_predict_without_fit), I only added a test for 2-class data to check whether the 1st level output is expected. Do you think I should add more to test the major part of the algorithm?

why dont you make a test on a 2-level binary hierarchy of clusters. and show that 1st level and 2nd level are both correct. can make all splits really easy.

> proof of effectiveness: do you think my simulation results (simple 1D 2GM binary & non-binary splits) would suffice? Should I try to recreate a hierarchical clustering figure from the literature? I don't really see which one in the reference paper would be a good choice...

(that) reference paper wont help us because it's for graphs. your simulation results are good. Maybe make one jupyter notebook and host it on github somewhere. make the simulation settings and results really clear like you have been doing. This will be great to have but a lower priority than getting the code in shape. Also consider having it be whatever simultations or results will go in the paper. 

> review: probably b/c I changed the location of the py file, it does not show any changes I made to the algorithm. How can I make it easier for you to review? Do you want me maybe to add the new file in the old (wrong) location just for now?

What you have now is fine, but if there were any unresolved comments/questions from the old file, please recomment as appropriate. 

> will add a tutorial next week

Sounds great. I have in mind what I think it should look like, so lets talk.

File: graspologic/cluster/rclust.py
Comment (tliu68) [inline graspologic/cluster/rclust.py:96]: > bdpedigo: not sure whether to call clusters or components because kmeans it would be clusters. we can discuss

File: graspologic/cluster/rclust.py
Comment (tliu68) [inline graspologic/cluster/rclust.py:306]: > bdpedigo: predict should not refit the model to the new data that was passed, should only run the prediction step. so will probably want to separate the predict functionality into this function or somewhere else, and then call it from fit_predict

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [reply graspologic/cluster/rclust.py:96]: let's ignore for now

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [reply graspologic/cluster/rclust.py:171]: apologies if this seems nitpicky but in general models don't store the data. it could be massive. and I think we can circumvent using the same recursive structure for `predict`

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [inline graspologic/cluster/rclust.py:215]: i said this above already, but `X` can just be passed in instead

Comment (bdpedigo) [reply graspologic/cluster/rclust.py:215]: furthermore, the whole point of `_cluster_and_decide` was to be a function (not a method part of a class) that could be used by this class and the HSBM class. So eventually will need to operate in that way anyway.

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [inline graspologic/cluster/rclust.py:131]: later in the code I never see you specify a value for these if they were passed as `None`. what happens then?

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [inline graspologic/cluster/rclust.py:50]: I don't actually understand what I'm supposed to pass for this as the user, what are the allowable strings I can use?

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [inline graspologic/cluster/rclust.py:199]: encourage you to use one of the tree traversal methods in anytree if possible. I bet there is a way to use those instead of this. maybe worth looking into at least?

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [inline graspologic/cluster/rclust.py:303]: apologies for repeating a comment  I think I said something similar elsewhere) but I still don't see why either needs to be stored as an attribute, same logic I think as for the `predict` function

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [inline graspologic/cluster/rclust.py:97]: I don't think this one is documented

File: tests/cluster/test_rclust.py
Comment (bdpedigo) [inline tests/cluster/test_rclust.py:84]: this test is great. but it does not cover all of your use cases, namely: 
- the different clustering methods you say you support
- the different criteria/likelihood ratio/delta stuff you say you support
and combinations of the above - can reuse the test you wrote for these other cases (there are lots of examples of this kinda thing in the tests folders).

File: tests/cluster/test_rclust.py
Comment (bdpedigo) [inline tests/cluster/test_rclust.py:82]: another test we want: predict works for non-fit data from the same distribution

**Review by bdpedigo — CHANGES_REQUESTED** [review]
hey @tliu68, gave some comments! I think they are all minor besides the structure of `_cluster_and_decide()` and the fact that you're storing data and other attributes that I don't see a reason to store. I think my comments here are very similar to those on `predict`, I think there is a fully recursive way to do this that will make the code shorter and avoid having to save big attributes (like the data itself).

Let me know if you have any questions or if I can help with writing anything for this!

**Comment by bdpedigo** [comment]
@tliu68 I am also unsure why the docs build is failing, if it keeps failing after you push again we can discuss

**Comment by bdpedigo** [comment]
also, unsure what to name this class officially - other common clustering stuff like Agglomerative is also recursive. We also probably don't want to call it Hierarchical because Agglomerative is also hierarchical. The only fully descriptive name I can think of is HierarchicalDivisiveCluster where Divisive basically means "top-down". But that seems like a bad name? so I'm stuck

File: graspologic/cluster/rclust.py
Comment (tliu68) [reply graspologic/cluster/rclust.py]: I didn't change this part cause I think maybe we want to do input checking of "level" (whether positive integer or not) before calling `predict` and check whether it's in the right range after calling `predict`

File: graspologic/cluster/rclust.py
Comment (tliu68) [reply graspologic/cluster/rclust.py]: I actually took out `selection_criteria` cause I think users can just specify that in `cluster_kws` if needed. If they didn't specify "AIC" for "gmm", it would be "BIC"; while there is no choice for "kmeans"

File: graspologic/cluster/rclust.py
Comment (tliu68) [reply graspologic/cluster/rclust.py]: i just added some explanation saying that if None, no such restriction on `delta-criter` applied. Also, after some experimentation, I could not find a proper case to show the effect of setting a significance threshold for likelihood_ratio test on improving the results. It seems that even when "BIC" improves a lot (several hundreds), the log_likehood changes very little so the p-value was really large and I couldn't get it pass the test. I think I'm gonna put a pin on this and do more research and add a working version later.

**Comment by tliu68** [comment]
> also, unsure what to name this class officially - other common clustering stuff like Agglomerative is also recursive. We also probably don't want to call it Hierarchical because Agglomerative is also hierarchical. The only fully descriptive name I can think of is HierarchicalDivisiveCluster where Divisive basically means "top-down". But that seems like a bad name? so I'm stuck

DivisiveClustering (sounds comparable to AgglomerativeClustering)?

File: graspologic/cluster/rclust.py
Comment (tliu68) [reply graspologic/cluster/rclust.py:95]: not sure whether I need to add more specific input checking for cluster_kws. For example, raise an error if someone tries to set linkage="ward" when choosing "kmeans" or completely random stuff.

File: graspologic/cluster/rclust.py
Comment (tliu68) [inline graspologic/cluster/rclust.py:210]: @bdpedigo Something I forgot to ask while ago: the min_comp is always 2 for `KMeansCluster`, so that clustering only stops if too few samples are left or current level is too deep. Do you think that's a problem because of inconsistency w/ gmm?

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [reply graspologic/cluster/rclust.py]: @tliu68 this was just a question. a little confused why it is only for first split

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [reply graspologic/cluster/rclust.py]: even better!

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [reply graspologic/cluster/rclust.py]: sounds great to me

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [reply graspologic/cluster/rclust.py:210]: @tliu68 can you explain what you mean by min_comp is always 2? I know that is the default. But KMeans should still run for k=1 (I think, based on a quick look at the code, i do know it's a bit silly). are you talking about the fact that silhouette_score doesnt make sense for k=1?

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [reply graspologic/cluster/rclust.py:95]: too complicated, let's leave that to the user to know what the models can accept

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [inline graspologic/cluster/rclust.py:57]: is None the same as 0?

**Comment by bdpedigo** [comment]
> > also, unsure what to name this class officially - other common clustering stuff like Agglomerative is also recursive. We also probably don't want to call it Hierarchical because Agglomerative is also hierarchical. The only fully descriptive name I can think of is HierarchicalDivisiveCluster where Divisive basically means "top-down". But that seems like a bad name? so I'm stuck
> 
> DivisiveClustering (sounds comparable to AgglomerativeClustering)?

Sure!

File: graspologic/cluster/rclust.py
Comment (tliu68) [reply graspologic/cluster/rclust.py:201]: Yes, I was talking about since silhouette_score is not defined for k=1, kmeans won't decide not to split b/c the current model has the best silhouette_score compared to models of more clusters.

File: graspologic/cluster/rclust.py
Comment (tliu68) [reply graspologic/cluster/rclust.py:53]: The way I imagine it should work is if the user wants to apply this constraint, they would specify a positive value, if no input is provided, it won't run. In the current code, 0 is not considered as a valid input. So I think right now 0 is not treated as None.

Comment (bdpedigo) [reply graspologic/cluster/rclust.py:53]: i thought it was doing "split if score_2 > score_1 + delta_crier" or the other way around, or something close to that, so I thought 0 would be a valid input. if I'm wrong, can you tell me what it's actually doing, and what values are allowed?

Comment (tliu68) [reply graspologic/cluster/rclust.py:53]: @bdpedigo I think we are talking about the same thing (this [line](https://github.com/microsoft/graspologic/blob/d5adacd2d95b3675e04924341ed088fb6596d9b0/graspologic/cluster/rclust.py#L197) is the same as your inequality). I was just thinking that if someone wants to make it harsher to accept a model of more clusters, they would input some value strictly greater than 0 (since if =0 it would be the same as what agmm does); None means don't do anything extra. But you are right 0 is not different from None here. I'll allow all non-negative values and just set 0 as default. Guess that's less confusing.

File: graspologic/cluster/rclust.py
Comment (tliu68) [inline graspologic/cluster/rclust.py:15]: @bdpedigo do you think adding "embed_kws" so that this function can be called by HSBMEstimator would be helpful?

Comment (bdpedigo) [reply graspologic/cluster/rclust.py:15]: yes but let's do that in next PR, in general, i try to avoid engineering for the future that explicitly (we may change our minds later)

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [reply graspologic/cluster/rclust.py:53]: sounds good. any scalar value seems fine then?

Comment (tliu68) [reply graspologic/cluster/rclust.py:53]: There is a nice summary of the rules of thumbs for both BIC & AIC in this [book](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C5&q=fabozzi+The+Basics+of+Financial+Econometrics%3A+Tools%2C+Concepts%2C+and+Asset+Management+Applications.&btnG=) that provides the original sources (specifically, on pages 401 & 403 [here](https://onlinelibrary.wiley.com/doi/pdf/10.1002/9781118856406.app5)). For both criteria, a difference of 10 seems to be strong enough (also what I used in the test).

File: graspologic/cluster/rclust.py
Comment (dwaynepryce) [inline graspologic/cluster/rclust.py:1]: Can we rename this file `recursive_cluster.py`?  Makes it easier to find it later if we name it that way off the top :)

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [inline graspologic/cluster/rclust.py:319]: @bdpedigo this part is unclear to me, but i dont have a clear idea of what to suggest, reminder to revisit

File: graspologic/cluster/rclust.py
Comment (bdpedigo) [reply graspologic/cluster/rclust.py:1]: we were debating renaming the class `DivisiveCluster` I think there is a comment about that somewhere, but happy to have the file name be more descriptive

Comment (bdpedigo) [reply graspologic/cluster/rclust.py:1]: @tliu68 can you make this change? And maybe we go with DivisiveCluster for now? Or DivisiveHierarchicalCluster if we wanna be overly verbose.

Comment (tliu68) [reply graspologic/cluster/rclust.py:1]: Will do!

Comment (dwaynepryce) [reply graspologic/cluster/rclust.py:1]: `divisive_cluster.py` and `DivisiveCluster` sound like fine names to me!

**Comment by bdpedigo** [comment]
hey @dwaynepryce i know you said this PR looked good but just wondering if we can get explicit approval in the next day or so so we can merge - i'll be giving it another quick look today but I don't think much will change.

**Review by dwaynepryce — CHANGES_REQUESTED** [review]
Still just need the `rclust.py` file to get named something like `divisive_cluster.py` so it's easy to find

**Comment by bdpedigo** [comment]
> Still just need the `rclust.py` file to get named something like `divisive_cluster.py` so it's easy to find

Done

**Comment by tliu68** [comment]
> Still just need the `rclust.py` file to get named something like `divisive_cluster.py` so it's easy to find

Sorry I was just afraid that renaming the file would make it hard to keep track of the changes and/or comments so was saving it for later.

File: tests/cluster/test_rclust.py
Comment (dwaynepryce) [inline tests/cluster/test_rclust.py:1]: Don't forget to rename me too!

Comment (tliu68) [reply tests/cluster/test_rclust.py]: Sure!

**Review by dwaynepryce — APPROVED** [review]

**Comment by dwaynepryce** [comment]
The conflicts are probably that I alphabetized the deps list

File: graspologic/cluster/divisive_cluster.py
Comment (bdpedigo) [inline graspologic/cluster/divisive_cluster.py:53]: @tliu68 can you confirm/deny the non-negative part being necessary here, and make any appropriate changes

File: graspologic/cluster/divisive_cluster.py
Comment (bdpedigo) [inline graspologic/cluster/divisive_cluster.py:153]: @tliu68 can you explain? when will they all be 0 vector?

Comment (bdpedigo) [reply graspologic/cluster/divisive_cluster.py:160]: part of why i bring this up - I am wondering whether we have some confusing behavior with regard to the leaf nodes. i.e. some leaf nodes could have no `.model_` and others could have a 1-cluster `.model_`. Is that what causes the thing you have to account for above?

**Comment by bdpedigo** [comment]
Some comments that we may want to discuss for the future:
- I think there should be the possibility of `delta_criter < 0` e.g. do less conservative splits. I have found this useful in practice. but right now this would break your code I think to simply allow this, so let's table the discussion for after this PR.
- had another comment about leaf node behaviors that we may want to save for after this PR? unsure. my question is, should we be consistent about how leaves are treated, as in, should they always/never have a `model_`? may make our lives easier to pick a standard.

**Review by bdpedigo — APPROVED** [review]

**Comment by bdpedigo** [comment]
@tliu68 I had a few minor comments and changes, wanted to give you the change to have a look tomorrow. but I am ready to merge whenever you are! Then we can move on to the tutorial? And then the relabeling?

Really awesome work, excited to have this. Also, congrats, this will be your first contribution to graspologic 🎉 🎉 🎉

File: graspologic/cluster/divisive_cluster.py
Comment (tliu68) [reply graspologic/cluster/divisive_cluster.py:164]: 1. at line 235 in `_fit` (`child_labels = dc._fit(new_X)`), if the prediction on node `dc` is a 0 vector, `child_labels` here is a 0 vector. If all nodes at the lowest level, i.e., all nodes with `depth = root.height`, have predictions of a 0 vector, the last column of the labels will be all 0s. 
Alternatively, we could add an if condition like `if np.max(child_labels) != 0` before the while loop next line to prevent adding this 0 vector to current `labels`. 
I honestly don't know which one is better code-wise or probably there is another better way. Both seem to be minor adjustment. I kind of think of this step as "cleaning up" the label output, so went w/ the former one to get rid of the redundant column in the end. I'm happy to do it differently if it seems weird to you.

1. Yes, not all nodes have `.model_` currently. The ones that did not pass the if conditions on lines 230-234 won't have it. We could add a sudo pred of 0 vector before the if conditions like I did at line 198
(not sure if those line #s are hard to find but I couldn't find a way to reference them, sorry about that...)

**Comment by tliu68** [comment]
> @tliu68 I had a few minor comments and changes, wanted to give you the change to have a look tomorrow. but I am ready to merge whenever you are! Then we can move on to the tutorial? And then the relabeling?
> 
> Really awesome work, excited to have this. Also, congrats, this will be your first contribution to graspologic 🎉 🎉 🎉

@bdpedigo They all look great to me! We will address the other comments in later PRs as you said. Thank you! And I'm proud of the work we did together and grateful for your help!

File: graspologic/cluster/divisive_cluster.py
Comment (bdpedigo) [inline graspologic/cluster/divisive_cluster.py:225]: @tliu68 does this need to be max_level - 1?

**Pull request merged**
